### ISSUE TITLE
can´t delete several pics at once

### ISSUE THREAD

**Issue by g6094199** [deltachat/deltachat-desktop#2892]
- Operating System (Linux/Mac/Windows/iOS/Android): Ubuntu Budgie 22.04
- Delta Chat Version: 1.32
- Expected behavior: possibility to pick several pics to take action on them, e.g. delete them at once
- Actual behavior: e.g .you can only delete one pic at a time, which is annoying when having a bunch of pics to delete
- Steps to reproduce the problem: try to pick more than one picture
- Screenshots:
![Bildschirmfoto vom 2022-09-21 11-33-06](https://user-images.githubusercontent.com/37833326/191470766-7b1b9d6c-6587-4e96-8f05-72c29d0f5d46.png)

**Comment by Jikstra** [comment]
Yes multi select would be an important feature :)

**Comment by Simon-Laux** [comment]
@JackJeisly suggested supporting both Ctrl+Click and Shift+Click, I assume the latter is for selecting ranges and the former for manual multi selection

**Comment by r10s** [comment]
> suggested supporting both Ctrl+Click and Shift+Click, I assume the latter is for selecting ranges and the former for manual multi selection

yes. this is a 30+ years old convention, you can check that eg. in windows explorer, macos finder and probably also most linux file managers. i miss that every day in delta chat desktop, for both, the chat-list and the message-list :)

**Comment by g6094199** [comment]
additionally for the desktop app: mark a message as mentioned above and delete it with the "del" (entf) key should also be possible.

**Comment by RasheedAZ** [comment]
Not sure if I should open a separate issue, but I'm having similar troubles deleting multiple chats. The only way possible appears to be right click and select "delete chat".

I have hundreds of chats that I need to delete :(

Already tried shift/ctr + select and also deleting with the delete key... none of them worked.

**Comment by farooqkz** [comment]
> Not sure if I should open a separate issue, but I'm having similar troubles deleting multiple chats. The only way possible appears to be right click and select "delete chat".
> 
> I have hundreds of chats that I need to delete :(
> 
> Already tried shift/ctrl + select and also deleting with the delete key... none of them worked.

You should open another issue.

**Comment by WofWca** [comment]
> supporting both Ctrl+Click and Shift+Click

For reference, check out ARIA guide, ~~[the Grid pattern](https://www.w3.org/WAI/ARIA/apg/patterns/grid/#gridNav_inside)~~ [the Listbox pattern](https://www.w3.org/WAI/ARIA/apg/patterns/listbox/):
> cell selection. If it did, though, the following keys are commonly used for these functions. 

This will probably need to be built on top of [improvement: a11y: arrow-key navigation for messages](https://github.com/deltachat/deltachat-desktop/pull/4294)

**Comment by WofWca** [comment]
I've been thinking about how multi-select would interact with our `useRovingTabindex` that we have for messages, chats, etc.
I'm getting an impression that the multi-select thing and the focus thing can be pretty much completely separate.
Basically with multi-select all we need to watch is item clicks, and whether Shift or Ctrl is pressed. Unless I'm missing something.

So, so far my vision is to have a generic `MultiselectContext` or something like that, which can be used in all the places we need it in.

OTOH maybe it's time to use one of the numerous "React listbox" components from NPM (e.g. https://react-spectrum.adobe.com/react-aria/ListBox.html). But I don't like how concrete / opinionated they are.

**Comment by Simon-Laux** [comment]
would also be cool if you could use multi select also with the keyboard, like press enter to select and move with arrow keys?

**Comment by WofWca** [comment]
Yes, that's what I'm saying. Make this thing "focus-based". And manage selection separately from focus.
"Enter" (or rather Space) would be a "click" event.

**Comment by r10s** [comment]
ftr, in the meantime, multi-select of chats is done, so this issue is only about multi-select of messages.

corresponding request in the forum: https://support.delta.chat/t/1169

**Comment by Raiden-GH** [comment]
I also think this is a very important feature! 
Perhaps it would also make sense to be able to select multiple contacts, like in DC-Android? 
That would be particularly convenient for deleting and maybe in the future for exporting vCards or sending a message to several people at once?

**Comment by WofWca** [comment]
Does anyone have an idea about what we should do against people attempting to e.g. forward 10 000 messages? With the current state https://github.com/deltachat/deltachat-desktop/pull/6268 it is possible. I commented about this in the MR:
> Consider adding a safeguard (maybe a confirmation dialog) if more than 100 or 1000 messages are selected.
Telegram simply won't let you select more than 100 messages.

**Comment by farooqkz** [comment]
> Does anyone have an idea about what we should do against people attempting to e.g. forward 10 000 messages? With the current state [#6268](https://github.com/deltachat/deltachat-desktop/pull/6268) it is possible. I commented about this in the MR:
> 
> > Consider adding a safeguard (maybe a confirmation dialog) if more than 100 or 1000 messages are selected.
> > Telegram simply won't let you select more than 100 messages.

Put a limit like 100. But also put a hard way to disable the limit. Like clicking on the build number 10 times :)

The goal is only people who really know what they are doing, and remove the limit.

**Comment by WofWca** [comment]
In reply to https://github.com/deltachat/deltachat-desktop/pull/6366#pullrequestreview-4264478271.

> but i really doubt that is needed, we have multi-selection on other platforms since years.

But on those platforms there is no contiguous selection, so one can't relatively easily select 99999 messages. On Desktop (with https://github.com/deltachat/deltachat-desktop/pull/6268) this is quite possible, using search to jump far back in chat history.
Do you think that displaying the selected count is enough of a guardrail still?

**Comment by r10s** [comment]
> Do you think that displaying the selected count is enough of a guardrail still?

if the issue is not about evil user, but about accidental taps and confused user, having the selected count always clearly in view is the more critical fix, yes [^1]

and if we assume that an additional dialog only saying "129 selected" - then maybe better improve the existing dialog by adding the impact there as well? so "Forward 129 messages to CHAT?" - that would also be an improvement for android/iOS

to protect more against evil user and making things not that easy for trolls, a hard limit, same as on telegram might be helpful, in case we regard other things as insufficient (there is a limit about how much can be sent, there are other ways to spam, maybe easier ones) (the discussion already started about how to disable that, however, is out of scope, we need some cut here)

[^1]: for accidental taps, it is also not not _that_ easy to select thousands of messages on desktop currently - eg. you cannot use the scrollbar to scroll to the first one. you would need to go to the first message by other means, search or so. maybe there is a shortcut. unless i missed smth, one really needs to want to do that.
that is similar to iOS, it is also not _that_ easy to select lots of messages - but doable: there you can scroll all the way up, and with two-finger gesture select lots of messages, while iOS accelerates ...

**Comment by WofWca** [comment]
> evil user

I'd say "a curious user" is a more common species. "Wait, would it let me forward over 9000 messages??"
But I got it.

**Comment by r10s** [comment]
> I'd say "a curious user" is a more common species. "Wait, would it let me forward over 9000 messages??"

as said, a hard limit may makes sense as well.

however, to move forward, i added a new string at https://github.com/deltachat/deltachat-android/pull/4427 , that allows to show the impact in the confirmation dialog. that, plus showing the number of items selected (which is super-useful als for deletion or whatever is added later on), should make a difference for the initial idea of "confused user"

**Comment by WofWca** [comment]
Remaining stuff:
- [ ] Improve the "just want to select text with Shift + Click" situation (comment):
  https://github.com/deltachat/deltachat-desktop/pull/6457#issuecomment-4681003204
- [ ] From https://github.com/deltachat/deltachat-desktop/pull/6268#issuecomment-4421606119
  - [ ] when tapping the "..." beside a message (instead of right tap), everything gets visually deselected - but it seems still to work. "..." should not deselect messages.
    https://github.com/deltachat/deltachat-desktop/pull/6682
  
  - [ ] the "emoji" button should be hidden in multi-select mode, it does not work, or it is unclear how it should work. just hide
- [ ] Add more actions to the context menu.
